File: backend/api/web/models/sample_input.py
"""APIs for SampleInput"""
from typing import Any, Dict, List, Optional

from modules.types import PMObject


class SampleInputInstance(PMObject):
    id: int
    created_at: str
    uuid: str

    name: Optional[str] = None
    content: Dict[str, Any]
    input_keys: Optional[List[str]] = None
    online: bool

    project_uuid: str
    function_model_uuid: Optional[str] = None
    ground_truth: Optional[str] = None


class CreateSampleInputBody(PMObject):
    project_uuid: str
    name: Optional[str] = None
    content: Dict[str, Any]
    function_model_uuid: Optional[str] = None
    input_keys: List[str]
    ground_truth: Optional[str] = None


class UpdateSampleInputBody(PMObject):
    name: Optional[str] = None
    content: Dict[str, Any]
    ground_truth: Optional[str] = None


class CreateDatasetBody(PMObject):
    project_uuid: str
    name: str
    description: Optional[str] = None
    function_model_uuid: Optional[str] = None


class UpdateDatasetBody(PMObject):
    name: Optional[str] = None
    description: Optional[str] = None


class DatasetInstance(PMObject):
    uuid: str
    name: str
    description: Optional[str] = None
    project_uuid: str
    eval_metric_uuid: Optional[str] = None
    function_model_uuid: Optional[str] = None


class DatasetSampleInputsCountInstance(PMObject):
    dataset_uuid: str
    count: int


class DatasetWithEvalMetricFunctionModelInstance(PMObject):
    dataset_uuid: str
    dataset_name: str
    dataset_description: Optional[str] = None
    eval_metric_uuid: str
    eval_metric_name: str
    eval_metric_description: Optional[str] = None
    function_model_uuid: Optional[str] = None
    function_model_name: Optional[str] = None


class CreateSampleInputForDatasetBody(PMObject):
    name: Optional[str] = None
    content: Dict[str, Any]
    input_keys: List[str]

    ground_truth: Optional[str] = None

File: frontend/src/types/SampleInput.ts
// Interfaces for SampleInput retrieval processes

/**
 * Interface for request to create a SampleInput.
 */
export interface CreateSampleInputRequest {
  project_uuid: string;
  name?: string | null;
  input_keys: Array<string>;
  content: Record<string, any>;
  function_model_uuid?: string | null;
  ground_truth?: string | null;
}

/**
 * Interface for request to update a SampleInput.
 */
export interface UpdateSampleInputRequest {
  uuid: string;
  name?: string | null;
  content: Record<string, any>;
  ground_truth?: string | null;
}

/**
 * Interface for request to create a Dataset.
 */
export interface CreateDatasetRequest {
  project_uuid: string;
  name: string;
  description?: string | null;
  function_model_uuid?: string | null;
}

/**
 * Interface for request to update a Dataset.
 */
export interface UpdateDatasetRequest {
  uuid: string;
  name?: string | null;
  description?: string | null;
}

/**
 * Interface for request to post multiple SampleInputs to an existing Dataset.
 */
export interface PostDatasetSampleInputsRequest {
  dataset_uuid: string;
  body: Array<{
    name?: string | null;
    input_keys: Array<string>;
    content: Record<string, any>;
    ground_truth?: string | null;
  }>;
}

/**
 * Interface for request to read Project's SampleInputs.
 */
export interface ReadProjectSampleInputsRequest {
  project_uuid: string;
}

/**
 * Interface for request to read a FunctionModel's SampleInputs.
 */
export interface ReadFunctionModelSampleInputsRequest {
  function_model_uuid: string;
}

/**
 * Interface for request to read a Dataset's SampleInputs.
 */
export interface ReadDatasetSampleInputsRequest {
  dataset_uuid: string;
  page: number;
  rows_per_page: number;
}

/**
 * Interface for request to read Dataset's SampleInputs count.
 */
export interface ReadDatasetSampleInputsCountRequest {
  dataset_uuid: string;
}

/**
 * Interface for response of Dataset's SampleInputs count.
 */
export interface ReadDatasetSampleInputsCountResponse {
  dataset_uuid: string;
  count: number;
}

/**
 * Interface for request to read a FunctionModel's Datasets.
 */
export interface ReadFunctionModelDatasetsRequest {
  function_model_uuid: string;
}

/**
 * Interface for request to delete a SampleInput.
 */
export interface DeleteSampleInputRequest {
  sample_input_uuid: string;
}

/**
 * Interface for datset with eval metric.
 */
export interface DatasetWithEvalMetric {
  id: number;
  dataset_uuid: string;
  dataset_name: string;
  dataset_description?: string | null;
  eval_metric_name: string;
  eval_metric_uuid: string;
  eval_metric_description?: string | null;
  function_model_uuid?: string | null;
  function_model_name?: string | null;
}
/**
 * General interface for SampleInput.
 */
export interface SampleInput {
  id: number;
  created_at: string;
  uuid: string;
  name: string;
  content: Record<string, any>;
  input_keys: Array<string>;
  online: true;
  project_uuid: string;
  function_model_uuid: string;
  ground_truth?: string | null;
}

/**
 * General interface for Dataset.
 */
export interface Dataset {
  id: number;
  uuid: string;
  name: string;
  description?: string | null;
  project_uuid: string;
  function_model_uuid?: string | null;
  eval_metric_uuid?: string | null;
}
